HDDS-5934. Add OM NodeID to VERSION file.

fapifta opened the pull request:
## What changes were proposed in this pull request?
OM uses its NodeID to generate the RaftPeerId. As such, the OM NodeId should be immutable. This Jira proposes to add the OM NodeId to its VERSION file. On restarts, the OM NodeID should be compared with the nodeId in the VERSION file to verify that it has not been updated.

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-5934

## How was this patch tested?
Added unit test

fapifta commented:
This is the same as #2800 with resolved conflicts.

errose28 commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:
```diff
@@ -459,6 +459,7 @@ private OzoneManager(OzoneConfiguration conf, StartupOption startupOption)
     this.omNodeDetails = omhaNodeDetails.getLocalNodeDetails();
 
     omStorage = new OMStorage(conf);
+    omStorage.setOmNodeId(omNodeDetails.getNodeId(), true);
```
Do we need this line and the `persistToDisk` flag? The node ID should have been persisted during init, so the constructor can just load it. The other fields in `OMStorage` don't have an extra `persistToDisk` boolean parameter.

errose28 commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOzoneManagerConfiguration.java:
```diff
@@ -486,4 +487,27 @@ private String getOMAddrKeyWithSuffix(String serviceId, String nodeId) {
+   * Re-initializing OMStorage with new nodeId should fail.
+   */
+  @Test
+  public void testChangeOMNodeId() throws Exception {
```
Can we also test a restart and reinit scenario where the value is changed in the config to be different from the value in the VERSION file? This would more closely resemble the mistake a user would make that we are trying to catch.

errose28 reviewed (commented):
Thanks for resuming work on this @fapifta .

kerneltime commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OMStorage.java:
```diff
@@ -67,6 +68,38 @@ public void setOmId(String omId) throws IOException {
+   * @throws IOException if existing nodeId from VERSION file does not match
+   * the new nodeId.
+   */
+  public void setOmNodeId(String nodeId, boolean persistToDisk)
```
```suggestion
  public void setOmNodeIdOnlyOnce(String nodeId, boolean persistToDisk)
```

fapifta commented:
@errose28, @kerneltime, first of all thank you for the reviews, all these comments were pretty much useful, I first just copied what was present in #2800, and fixed the conflict with master.

Based on your comments, I realized a few things about what we try to achieve with different parts of the patch, and also about how this should be tested.

At the end of the day, I removed the test that was added by the original patch, and added some additional tests as well to cover the OMStorage class more.
I also have changed the API, so now there is one method that provides a way to set the OM Node ID at init, when the VERSION file is being initialized first, and added a convenience method for already existing installations where the VERSION file is already present without this value, along with the initial verification logic that checks if the node id is set in the VERSION file, then that is matching what we have got from the configuration.

This way we do not need the boolean parameter, and we can defined clearly the difference between the two case.

Please check the updated solution, and let me know what do you think!
Thank you!

errose28 commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OMStorage.java:
```diff
@@ -32,52 +32,178 @@
-  public static final String OM_CERT_SERIAL_ID = "omCertSerialId";
+  static final String ERROR_OM_IS_ALREADY_INITIALIZED =
+      "OM is already initialized.";
+  static final String ERROR_UNEXPECTED_OM_NODE_ID_TEMPLATE =
+      "OM NodeId: {} does not match existing nodeId from VERSION file: {}";
```
This is being passed to `String#format`, which uses `%` as the encoding, so the `{}` instances are not replaced. Intellij has a warning for this. `testValidateOrPersistOmNodeIdThrowsWithDifferentNodeIdAsInFile` makes the same mistake, and only checks for substring matches, so it passes.

fapifta replied:
Oh... my bad, I don't know what I was doing here :D Fixing it.

errose28 commented on hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/TestOMStorage.java:
```diff
@@ -17,76 +17,312 @@
+    return ret;
+  }
+
+  private OMStorage setupAPersistedVersionFile(OzoneConfiguration conf)
```
Looks like the return of this method is never used. Should it be removed or is there a use case for the return value that got missed?

fapifta replied:
First it was used, then I realized that an OMStorage object never gets initialized just if it is being created with a config that points to a directory that contains a proper VERSION file, after that the return value was not necessary anymore, but I did not realized this. Thank you for recognizing this, I am changing the return value to void.

errose28 reviewed (commented):
Thanks for taking the time to refactor and improve this area of the code @fapifta. I left some minor comments in line.

fapifta commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OMStorage.java:
```diff
@@ -32,52 +32,178 @@
+ * On top of what is defined in the base Storage class, this class adds
+ * functionality to hold Ozone Manager related data in its VERSION file.
+ * The additional values stored:
+ * - Ozone Manager ID - a UUID that identifies this Ozone Manager.
+ *                      The value can not be changed once initialized.
+ * - Ozone Manager Node Id - the node id defined for this Ozone manager in the
+ *                           configuration. The value can not be changed after
+ *                           it was set.
```
Yes you are right, I am updating this doc, with a note that this one is not used anymore. It is a required part of a protocol message OzoneManagerDetailsProto in HDDSProtos, so I am hesitant to remove it fully. We can do it but that should be part of a separate JIRA if we want to, however I think as this is a UUID even if we do not use it for anything at the moment, it is still an identifier of the OM, and does not require too much maintenance so I would not remove it yet.

fapifta commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:
```diff
@@ -459,6 +459,7 @@ private OzoneManager(OzoneConfiguration conf, StartupOption startupOption)
     this.omNodeDetails = omhaNodeDetails.getLocalNodeDetails();
 
     omStorage = new OMStorage(conf);
+    omStorage.validateOrPersistOmNodeId(omNodeDetails.getNodeId());
```
Comment on the sidenote:
As a possible solution we can bump the layout version, and add an UpgradeAction to handle the change in the VERSION file, however I am not sure whether this is an overkill for just adding this value. Probably it is not, as we are changing the metadata, but I am still a bit hesitant to go and do it without hearing your opinion on this :)

I thought adding a comment here is just redundant, as the doc of the method is pretty much trying to tell what it is doing, but anyway... why not :) I have added a code comment before the method call to explain, however if we decide to do this within the upgrade framework, this one will be meaningless anyway ;)

errose28 replied:
I can see 3 possible options:

1. Write the value in the constructor if it is not present, or validate the value if it is already there.
    - Simple approach
    - All installs get the value without a re-init.
    - May lead to code fragmentation between omInit and the constructor for all new fields added to the version file.
        - omInit must write all the values, but the constructor must also write any new values. Previously the constructor did not write to the version file.
2. Add a layout feature and upgrade action to write the new value to the version file. omInit writes the value to the version file, and OM constructor only validates the value.
    - More involved approach
    - All installs get the value without a re-init.
    - omInit and the constructor maintain their original jobs with respect to the version file, but some upgrade pieces must be added.
3. Write the value in omInit only. Existing clusters that do not re-init will not have the value persisted.
    - Simple approach
    - Only new installs or re-inited existing installs get it.
    - omInit and constructor maintain their original jobs with respect to the version file.

2 seems excessive IMO since there are no downgrade implications to the change. 3 may be annoying to debug in the future. Someone may change the node ID and this change will not catch it even if it is present. So after writing out the options like this, I think 1 is the best choice, which is what the PR is already doing.

After another look the javadoc on the method should be sufficient. If anyone is wondering why the constructor may need to modify the version file on startup they would probably go from the constructor to the method declaration and see the doc.

fapifta replied:
Ok, I think we agree on this then. I have added these options as a comment above the validateOrStore method.

fapifta replied:
Oh, one more thing...
I went ahead and tried to add an UpgradeAction, but once I started I quickly realized the following:
UpgradeActions are running pretty much after Ozone Manager was initialized almost fully, in the start() method, it seems to be pretty late to fail, and the impacts of moving the initialization of the value in the storage or the validation to that late are not clear, so it is at least risky.

errose28 commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:
```diff
@@ -454,6 +454,11 @@ private OzoneManager(OzoneConfiguration conf, StartupOption startupOption)
     this.omNodeDetails = omhaNodeDetails.getLocalNodeDetails();
 
     omStorage = new OMStorage(conf);
+    // validate if the stored one is the same we get in the OMNodeDetails
+    // if not, fails the start. If the VERSION file does not contain the node
+    // id yet, it stores it to the VERSION file and skips validation for the
+    // that case.
```
```suggestion
    // validate if the stored node ID is the same we get in the OMNodeDetails
    // if not, fails to start. If the VERSION file is from an older Ozone version and does not contain the node
    // ID yet, it stores it to the VERSION file and skips validation.
```

Minor suggestion here to clarify why the version file may not have the node ID and we need to write it. If you just read the code after this change, it looks like the node ID should show up on init and the constructor can assume it is already there. It is not obvious that the node ID used to not be written in init.

fapifta replied:
Based on the previous comment where you seemed to agree that the method apidoc is enough, I removed this part, and extended the apidoc a bit.

errose28 reviewed (commented):
LGTM with just a minor suggestion.

fapifta commented:
Hi @errose28 thank you for the review, I have made some adjustments to the doc and removed the comment, I think it will be good this way, let me know if you agree, and if you are ok with the patch, I going to commit it sometime tomorrow. @kerneltime if you want to take a second look, let me know, and I will wait for that ;)

errose28 reviewed (approved):
Thanks @fapifta LGTM.
